### ISSUE TITLE
implement `log2`, `log10`, `log256` (and possibly `ln`)

### ISSUE THREAD

**Issue by diwu1989** [holiman/uint256#123]
Can there be a natural log or base 2 log implemented?
Thanks

**Comment by A-ndy-git** [comment]
+1. Also been looking for this functionality for some time.

I'm unsure of the best method to implement this?

**Comment by elee1766** [comment]
there are a few different solidity libraries that implement logarithm. off the top of my head, the big name i know is oz, they have log2,log10,log256 in their math libraries: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/math/Math.sol 
log10 i think is used in their word -> ascii string function.

it may be desirable to replicate the results of one such library (not neccesarily oz's), since I think mirroring or performing on-chain operations is a common use case for this package.

**Comment by diwu1989** [comment]
+1 yes, that was my reasoning for requesting this, because I wanted to match some of the LOG math used in smart contracts

**Comment by elee1766** [comment]
> +1 yes, that was my reasoning for requesting this, because I wanted to match some of the LOG math used in smart contracts

happy to do some work on this. personally we use log2

which LOG math functions do you have immediate needs for? i can start with those and add more down the line

log2? log10? log256?

**Comment by diwu1989** [comment]
Log2 is enough for UniswapV3 tickmath:
https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/TickMath.sol

**Comment by holiman** [comment]
Is there any difference between `Log2()` and `BitLen()`? Because the latter exists already...

**Comment by karalabe** [comment]
> Is there any difference between `Log2()` and `BitLen()`? Because the latter exists already...

Yes, `BitLen` is `int(log2)`. But if you want to calculate `logB(N)`, you need `log2(N)/log2(B)` with full float representation, not truncated to int.

**Comment by karalabe** [comment]
FWIW, here's a set of log implementation for big.Int https://github.com/robpike/ivy/blob/master/value/log.go

**Comment by holiman** [comment]
Possible api-methods could be 
```
(z *Int) IntLog2( ... ) *Int
(z *Int) IntLog10( ... ) *Int

(z *Int) Log2( ... ) float64
(z *Int) Log10( ... ) float64
```
However, it would be simpler to just do 
```
(z *Int) IntLog2( ... ) *Int
(z *Int) IntLog10( ... ) *Int

(z *Int) Float64() float64

```

Then the caller can just do `math.Log2(myint.ToFloat64())` to deal in `float`s. The `math` package already has it all.

**Comment by elee1766** [comment]
@diwu1989 @holiman 

https://gfx.cafe/-/snippets/10

the log2 required in uniswap is just msb, so bitlen - 1

**Comment by holiman** [comment]
Yeah implementing `Log2` and `Log256` is pretty trivial, but implementing `Log10` is a bit more fun.

**Comment by holiman** [comment]
`Log10` implemented by #136
`Float64` implemented by #132 

If anyone wants to add `Log2` and `Log256`, it's welcome. Otherwise, I think most users can just use the existing `BitLen`. I'll close this as 'done'.
